src/core/library.c

```diff
@@ -505,6 +505,58 @@ QuicLibApplyLoadBalancingSetting(
         MsQuicLib.CidTotalLength);
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicLibrarySumPerfCounters(
+    _Out_writes_bytes_(BufferLength) uint8_t* Buffer,
+    _In_ uint32_t BufferLength
+    )
+{
+    const uint32_t CountersPerBuffer = BufferLength / sizeof(uint64_t);
+    uint64_t* const Counters = (uint64_t*)Buffer;
+    memcpy(Buffer, MsQuicLib.PerProc[0].PerfCounters, BufferLength);
+
+    for (uint32_t ProcIndex = 1; ProcIndex < MsQuicLib.ProcessorCount; ++ProcIndex) {
+        for (uint32_t CounterIndex = 0;
+            CounterIndex < CountersPerBuffer;
+            ++CounterIndex) {
+            Counters[CounterIndex] =
+                ((int64_t)Counters[CounterIndex]) +
+                (int64_t)MsQuicLib.PerProc[ProcIndex].PerfCounters[CounterIndex];
+        }
+    }
+
+    if (QUIC_PERF_COUNTER_CONN_ACTIVE < CountersPerBuffer &&
+        (int64_t)Counters[QUIC_PERF_COUNTER_CONN_ACTIVE] < 0) {
+        Counters[QUIC_PERF_COUNTER_CONN_ACTIVE] = 0;
+    }
+
+    if (QUIC_PERF_COUNTER_CONN_CONNECTED < CountersPerBuffer &&
+        (int64_t)Counters[QUIC_PERF_COUNTER_CONN_CONNECTED] < 0) {
+        Counters[QUIC_PERF_COUNTER_CONN_CONNECTED] = 0;
+    }
+
+    if (QUIC_PERF_COUNTER_PKTS_SUSPECTED_LOST < CountersPerBuffer &&
+        (int64_t)Counters[QUIC_PERF_COUNTER_PKTS_SUSPECTED_LOST] < 0) {
+        Counters[QUIC_PERF_COUNTER_PKTS_SUSPECTED_LOST] = 0;
+    }
+
+    if (QUIC_PERF_COUNTER_STRM_ACTIVE < CountersPerBuffer &&
+        (int64_t)Counters[QUIC_PERF_COUNTER_STRM_ACTIVE] < 0) {
+        Counters[QUIC_PERF_COUNTER_STRM_ACTIVE] = 0;
+    }
+
+    if (QUIC_PERF_COUNTER_CONN_QUEUE_DEPTH < CountersPerBuffer &&
+        (int64_t)Counters[QUIC_PERF_COUNTER_CONN_QUEUE_DEPTH] < 0) {
+        Counters[QUIC_PERF_COUNTER_CONN_QUEUE_DEPTH] = 0;
+    }
+
+    if (QUIC_PERF_COUNTER_OPER_QUEUE_DEPTH < CountersPerBuffer &&
+        (int64_t)Counters[QUIC_PERF_COUNTER_OPER_QUEUE_DEPTH] < 0) {
+        Counters[QUIC_PERF_COUNTER_OPER_QUEUE_DEPTH] = 0;
+    }
```

Comment:
I'm thinking it's still better to just have a big loop and not be type specific. It scales better.

Reply:
Well some of these counter might actually be unsigned values that large. We wouldn't want to zero those out.

Reply:
You think some of these counters will be greater than 2^63? That's [nearly 300 years](https://www.bing.com/search?q=2%5E63+nanoseconds+to+years&qs=n&form=QBRE&sp=-1&pq=2%5E63+nanoseconds+to+years&sc=5-25&sk=&cvid=F6D98D58B097428CAF2BAF6DF4AF2EB4) if the counter is incremented once every nanosecond...

Reply:
I was thinking that the quic stream data counters are 62 bits, and summing them up might overflow a signed value. But even with a 100Gb NIC, that would take 23 years. Point taken.